### PR TITLE
osd/OSDMap: map pgs with smaller batchs in calc_pg_upmaps

### PR DESCRIPTION
**Pull request by hjwsm1989**
We have a cluster pool with 32768 pgs and 400 osds, it costs 600 seconds when doing upmap
with '--upmap-max 32768 --upmap-deviation 0.01', which is pretty slow.
After adding some debug code, the time mostly spent on pg_to_up_acting_osds, the average
time for one pg_to_up_acting_osds is about 12us, so the whole pool's pg will cost 500ms each
time, we finally have 1429 pgs need to do upmap, so it cost about 600 seconds.
Withi this patch, it only spend 5 seconds to get job done.

Signed-off-by: huangjun <huangjun@xsky.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/osd/OSDMap.cc [hunk 1]
```diff
@@ -4247,51 +4247,50 @@ int OSDMap::calc_pg_upmaps(
...
-      else
+    else
 	pgs_by_osd.emplace(i.first, set<pg_t>());
-      ldout(cct, 20) << " osd." << i.first << " weight " << i.second
+    ldout(cct, 20) << " osd." << i.first << " weight " << i.second
 		     << " pgs " << pgs << dendl;
-    }
-
+  }
+  while (true) {
```
Comment (xiexingguo) [inline src/osd/OSDMap.cc]: Can you move the `while `loop a few lines down, just before  the ` // osd deviation` comment?

**Comment by xiexingguo** [comment]
one nit, otherwise this looks quite reasonable to me

File: src/osd/OSDMap.cc (hunk 1 above)
Comment (hjwsm1989) [reply src/osd/OSDMap.cc]: Thanks, updated, please review.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-08-28_04:25:25-rados-wip-kefu-testing-2018-08-28-0811-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-08-29_07:26:03-rados-wip-kefu-testing-2018-08-28-0811-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-08-29_11:04:40-rados-wip-kefu-testing-2018-08-28-0811-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-08-29_13:22:00-rados-wip-kefu-testing-2018-08-28-0811-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-08-29_15:16:19-rados-wip-kefu-testing-2018-08-28-0811-distro-basic-mira/

fixes posted at #23681 and #23803

**Pull request merged**